### PR TITLE
[release/2.8] deprecate reference package, migrate to github.com/distribution/reference

### PR DESCRIPTION
**Pull request by thaJeztah**
- (adjusted) backport of https://github.com/distribution/distribution/pull/4031
- relates to https://github.com/distribution/distribution/issues/3587

This integrates the new module, which was extracted from this repository at commit b9b19409cf458dcb9e1253ff44ba75bd0620faa6;

    # install filter-repo (https://github.com/newren/git-filter-repo/blob/main/INSTALL.md)
    brew install git-filter-repo

    # create a temporary clone of docker
    cd ~/Projects
    git clone https://github.com/distribution/distribution.git reference
    cd reference

    # commit taken from
    git rev-parse --verify HEAD
    b9b19409cf458dcb9e1253ff44ba75bd0620faa6

    # remove all code, except for general files, 'reference/', and rename to /
    git filter-repo \
      --path .github/workflows/codeql-analysis.yml \
      --path .github/workflows/fossa.yml \
      --path .golangci.yml \
      --path distribution-logo.svg \
      --path CODE-OF-CONDUCT.md \
      --path CONTRIBUTING.md \
      --path GOVERNANCE.md \
      --path README.md \
      --path LICENSE \
      --path MAINTAINERS \
      --path-glob 'reference/*.*' \
      --path-rename reference/:

    # initialize go.mod
    go mod init github.com/distribution/reference
    go mod tidy -go=1.20

This commit is based on 152af63ec5c569f074e9cf5d0e409d6928e034d8 in the main branch, but adjusted for the 2.8 branch, with some differences:

- the Sort functions have not been kept, as they were not part of the v2 package, and introduced in 1052518d9f962372694e004dba46d2036fb23101 (https://github.com/distribution/distribution/pull/3777)
- the ParseAnyReferenceWithSet and ShortIdentifierRegexp were kept (but deprecated) as removing happened in 6d4f62d7fdfa25bd4bb42a18995c50aeededc0d6 (https://github.com/distribution/distribution/pull/3774), which is not in the 2.8 branch.

### CODE REVIEW

File: reference/normalize_deprecated.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+func (d digestReference) Digest() digest.Digest {
+	return digest.Digest(d)
+}
+
+// ParseAnyReferenceWithSet parses a reference string as a possible short
+// identifier to be matched in a digest set, a full digest, or familiar name.
+//
+// Deprecated: support for short-identifiers is deprecated, and will be removed in v3.
+func ParseAnyReferenceWithSet(ref string, ds *digestset.Set) (Reference, error) {
```
Comment (thaJeztah) [inline reference/normalize_deprecated.go:79]: This was a bit unfortunate that it was still in the 2.8 branch, but I thought it was a good opportunity to add the Deprecated message here (although there don't appear to be any external consumers; see https://github.com/distribution/distribution/pull/3774)

File: reference/helpers_deprecated.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Deprecated: use [reference.IsNameOnly].
+func IsNameOnly(ref reference.Named) bool {
+	return reference.IsNameOnly(ref)
+}
+
+// FamiliarName returns the familiar name string
+// for the given named, familiarizing if needed.
+//
+// Deprecated: use [reference.FamiliarName].
```
Comment (thaJeztah) [inline reference/helpers_deprecated.go:15]: All functions / types were marked as deprecated, but continue to be functional, so consumers _can_ decide to continue using them, but the deprecation message should hopefully make them start preparing to move over to the new package (before switching to distribution v3 once a release for that that arrives)

**Comment by thaJeztah** [comment]
One small fix needed, looks like;

```
level=warning msg="[runner] Can't run linter goanalysis_metalinter: buildir: failed to load package : could not load export data: no export data for \"github.com/distribution/distribution/digestset\""
level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: buildir: failed to load package : could not load export data: no export data for \"github.com/distribution/distribution/digestset\"\n\n"
```

**Comment by thaJeztah** [comment]
All green now; issue was because the version of go-digest in this branch was a pre-v1.0.0 release, which didn't yet have the `digestset` package. I opened a follow-up to consider as well;

- https://github.com/distribution/distribution/pull/4064

**Review by milosgajdos — APPROVED** [review]
Noice!

**Pull request merged**